### PR TITLE
test(linter/rules-of-hooks): add edge cases to test suite

### PR DESCRIPTION
**Pull request by camc314**

### CODE REVIEW

**Comment by camc314** [comment]
* **#18099** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/18099?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/18099?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=18099).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/oxc_linter/src/rules/react/rules_of_hooks.rs [hunk 1]
```diff
@@ -1064,7 +1064,8 @@ fn test() {
     // https://github.com/oxc-project/oxc/issues/6651
     r"const MyComponent = makeComponent(() => { useHook(); });",
     r"const MyComponent2 = makeComponent(function () { useHook(); });",
-    r"const MyComponent4 = makeComponent(function InnerComponent() { useHook(); });"
+    r"const MyComponent4 = makeComponent(function InnerComponent() { useHook(); });",
+    r"const Foo = hoc((props) => { if (props.cond) { const [_a, _b] = useState(false); } });"
```
[the remark this answers is not in the document]
Comment (camc314) [reply crates/oxc_linter/src/rules/react/rules_of_hooks.rs:1068]: although strange, this follows the same behaviour as the original eslint plugin
